power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -21,112 +25,249 @@ namespace detail = power_grid_model::optimizer::detail;
 
 using TrafoGraphIdx = Idx;
 using EdgeWeight = int64_t;
-using WeightedTrafo = std::pair<Idx2D, EdgeWeight>;
-using WeightedTrafoList = std::vector<WeightedTrafo>;
-using RankedTransformerGroups = std::vector<std::vector<Idx2D>>;
 constexpr auto infty = std::numeric_limits<Idx>::max();
 
 struct TrafoGraphVertex {
-    bool is_source{}; // is_source = true if the vertex is a source
+    bool is_source{};
 };
 
 struct TrafoGraphEdge {
-    Idx2D pos{};
+    bool is_trafo{false};
+    Idx2D from_to{};
     EdgeWeight weight{};
+
+    bool operator==(const TrafoGraphEdge& other) const {
+        return is_trafo == other.is_trafo && from_to == other.from_to && weight == other.weight;
+    }
+
+    bool operator<(const TrafoGraphEdge& other) const {
+        if (weight != other.weight) {
+            return weight < other.weight;
+        }
+        if (from_to.group != other.from_to.group) {
+            return from_to.group < other.from_to.group;
+        }
+        return from_to.pos < other.from_to.pos;
+    }
```

Comment:
please instead, define the spaceship operator `<=>`. this was added in C++20 precisely because it is otherwise so verbose to write. this is a really good writeup on that: https://indii.org/blog/overloading-the-spaceship-operator/

i believe that in this case the syntax is something like

```suggestion
    // weak ordering because we don't explicitly check for is_trafo
    std::weak_ordering operator<=>(const TrafoGraphEdge& other) const {
        if (auto cmp = weight <=> other.weight; cmp != 0) {
            return cmp;
        }
        if (auto cmp = from_to.group <=> other.from_to.group; cmp != 0) {
            return cmp;
        }
        return from_to.pos <=> other.from_to.pos;
    }
```

we can also decide to overload this operator<=> for `Idx2D` and then it becomes even shorter. i think that makes sense cause Idx2D is basically ordered due to the implied meaning of `group` and `pos`

```suggestion
    // weak ordering because we don't explicitly check for is_trafo
    std::weak_ordering operator<=>(const TrafoGraphEdge& other) const {
        if (auto cmp = weight <=> other.weight; cmp != 0) {
            return cmp;
        }
        return from_to <=> other.from_to;
    }
```

fun fact, if you declare `weight` before `from_to`, then it becomes even easier, because the first argument is compared first by default (implying strong ordering):

```suggestion
    friend auto operator<=>(const TrafoGraphEdge& a, const TrafoGraphEdge& b) = default;
```

i believe that that is also sufficient for `Idx2D`:
```cpp
    friend auto operator<=>(const Idx2D& a, const Idx2D& b) = default;
```

Reply:
The comparison operators are actually necessary evil in test cases as in production a sorted list within group is not necessary. If shuffling the lines a bit makes things so much simpler, why nut?